Allow to prove the non-existence of keys

hrxi opened the pull request:
This is done by including the last existing node on the path to the missing key. In order to make verifying cheap, the mapping from the missing key to the last existing node's key is included in the proof.

Includes the commits from #1455, because it would conflict otherwise.

hrxi committed:
trie: Fix node counting if we remove a non-existing key

hrxi committed:
`todo!()` → `unreachable!()`

The `Stream` implementation of `wasm_timer::Interval` cannot return
`Poll:Ready(None)`.

hrxi committed:
Normalize some `Trie` code

Name types the same way throughout the file, drop unneeded extra traits.

sisou commented on primitives/src/trie/trie_proof.rs:
```diff
@@ -123,32 +139,105 @@ impl TrieProof {
+            k1p.post_order_cmp(k2p).then_with(|| k1.post_order_cmp(k2))
+        });
+
+        eprintln!("{:?}", keys);
+        eprintln!(
+            "{:?}",
+            self.nodes.iter().map(|n| &n.key).collect::<Vec<_>>()
+        );
```
Debug output?

hrxi replied:
Correct, thank you. Fixed.

sisou commented on primitives/src/trie/trie_proof.rs:
```diff
@@ -123,32 +139,72 @@ impl TrieProof {
+                        cur_key = keys.next();
+                        cur_node = nodes.next();
+                    } else {
+                        eprintln!("checking {} against {} for negativity", k, n.key);
```
Still a remaining `eprintln`. Should be a `log::debug`?

hrxi replied:
Hmmm hmmm, not sure how I forgot that one… Removed and added a comment instead.

hrxi committed:
Allow to prove the non-existence of keys

This is done by including the last existing node on the path to the
missing key. In order to make verifying cheap, the mapping from the
missing key to the last existing node's key is included in the proof.

jsdanielh commented:
Rebasing it to get it merged
